apm-agent-plugins/apm-es-restclient-plugin/src/main/java/co/elastic/apm/es/restclient/ESRestClientInstrumentationHelper.java

```diff
@@ -0,0 +1,59 @@
+/*-
+ * #%L
+ * Elastic APM Java agent
+ * %%
+ * Copyright (C) 2018 Elastic and contributors
+ * %%
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * #L%
+ */
+package co.elastic.apm.es.restclient;
+
+import co.elastic.apm.bci.VisibleForAdvice;
+import co.elastic.apm.report.serialize.DslJsonSerializer;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nullable;
+import java.io.IOException;
+import java.io.InputStream;
+
+@VisibleForAdvice
+public class ESRestClientInstrumentationHelper {
+    private static final Logger LOGGER = LoggerFactory.getLogger(ESRestClientInstrumentationHelper.class);
+
+    private static ThreadLocal<byte[]> bodyReadBuffer = new ThreadLocal<>();
+
+    @Nullable
+    @VisibleForAdvice
+    public static String readRequestBody(InputStream bodyIS, String endpoint) throws IOException {
+        String body = null;
+        try {
+            byte[] data = bodyReadBuffer.get();
+            if (data == null) {
+                bodyReadBuffer.set(new byte[DslJsonSerializer.MAX_LONG_STRING_VALUE_LENGTH]);
```

Comment:
That limits the number of bytes as opposed to the number of characters. In case of non-ascii characters, there is a difference. Although, for simplicity's sake it probably still makes sense the way you do it. Even if it means there could be less than 10k chars, but it might be worth adding a comment about that.

Reply:
Well, that's the least of problems here- if the binary data was encoded using a charset other than the platform default, then I would mess it up when decoding it to a String. I wrote the proper handling for dealing with that, which was to get the content-type header from the request, parse it, see if it contains a charset and use it for decoding, but only then I realized that it will be messed up during JSON serialization anyway, so I dropped it for now.

Anyway, I guess the main issue here is size limit, not a specific number, so 10k has no special meaning and I believe it achieves its purpose.

Reply:
I'd assume that the ES RestClient uses a fixed encoding, probably UTF-8. Making that assumption very clear, for example by naming the method `readRequestBodyUtf8`, I think It's probably good enough to hard-code the encoding rather than to determine it at runtime.

> but only then I realized that it will be messed up during JSON serialization anyway

not quite sure I understand

Reply:
> I'd assume that the ES RestClient uses a fixed encoding, probably UTF-8

Yeah, you are probably right, I treated it as if it's a general-purpose HTTP client, but it doesn't really make sense